data/mongodb/v0/tasks.go

```diff
@@ -0,0 +1,280 @@
+		return nil, err
+	}
+
+	outputStruct := InsertOutput{
+		Status: "Successfully inserted document",
+	}
+
+	output, err := base.ConvertToStructpb(outputStruct)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func (e *execution) find(in *structpb.Struct) (*structpb.Struct, error) {
+	var inputStruct FindInput
+	err := base.ConvertFromStructpb(in, &inputStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	criteria := inputStruct.Criteria
+	limit := inputStruct.Limit
+	ctx := context.Background()
+
+	realCriteria := bson.M{}
+	for key, value := range criteria {
+		if value != nil {
+			realCriteria[key] = value
+		}
+	}
+
+	findOptions := options.Find()
+
+	if limit > 0 {
+		findOptions.SetLimit(int64(limit))
+	}
+
+	var cursor *mongo.Cursor
+	if len(criteria) == 0 {
+		// Find all documents with the specified options (including limit if set)
+		cursor, err = e.collectionClient.Find(ctx, realCriteria, findOptions)
+	} else {
+		// Define the projection to include only the specified fields
+		projection := bson.M{}
+		projection["_id"] = 0
+		for key := range criteria {
+			projection[key] = 1
+		}
+		findOptions.SetProjection(projection)
+
+		// Find all documents with the specified options (including limit if set)
+		cursor, err = e.collectionClient.Find(ctx, realCriteria, findOptions)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var documents []map[string]any
+	for cursor.Next(ctx) {
+		var row map[string]any
+		err := cursor.Decode(&row)
+		if err != nil {
+			return nil, err
+		}
+		documents = append(documents, row)
+	}
+
+	outputStruct := FindOutput{
+		Status:    "Successfully found documents",
+		Documents: documents,
+	}
+
+	output, err := base.ConvertToStructpb(outputStruct)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func (e *execution) update(in *structpb.Struct) (*structpb.Struct, error) {
+	var inputStruct UpdateInput
+	err := base.ConvertFromStructpb(in, &inputStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	criteria := inputStruct.Criteria
+	updateFields := inputStruct.Update
+	ctx := context.Background()
+
+	// Get the document to identify which fields need to be removed
+	var result map[string]interface{}
+	err = e.collectionClient.FindOne(ctx, criteria).Decode(&result)
+	if err != nil {
+		return nil, err
+	}
+
+	setFields := bson.M{}
+	unsetFields := bson.M{}
+
+	for key := range result {
+		// If the field in the existing document is not in the updateFields, it should be removed
+		if _, found := updateFields[key]; !found && key != "_id" {
+			unsetFields[key] = ""
+		}
+	}
+
+	for key, value := range updateFields {
+		setFields[key] = value
+	}
+
+	updateDoc := bson.M{}
+	if len(setFields) > 0 {
+		updateDoc["$set"] = setFields
+	}
+	if len(unsetFields) > 0 {
+		updateDoc["$unset"] = unsetFields
+	}
+
+	if len(updateDoc) == 0 {
+		return nil, fmt.Errorf("no valid update operations found")
+	}
+
+	_, err = e.collectionClient.UpdateMany(ctx, criteria, updateDoc)
```

Comment:
question
I saw in json schema, you set update for 1 data.
But why here is `UpdateMany`?

Reply:
@chuang8511 

> I saw in json schema, you set update for 1 data.

Sorry but I would like to know which json schema is it ? I always intend to use UpdateMany since it can be used for both single or multiple document. Field like `"city":"New York"` on criteria/filter may have multiple documents, and updated to specific structure. We could also make use of mongodb `_id` if we want one specific document combined with another field.

Reply:
@zidanehakim 
```
        "update-data": {
          "description": "The updated data to be applied to the documents",
          "instillAcceptFormats": [
            "semi-structured/*","structured/*","object"
          ],
          "instillUIOrder": 3,
          "instillUpstreamTypes": [
            "reference",
            "template",
            "value"
          ],
          "title": "Update"
        }
```

I thought it should be array:object, array:semi-structured/*, ... right?
I am not so sure if object contains array concept.

Reply:
@chuang8511

Oh, my intention is for instance if we have 3 documents with same city and different name,
`{"name":"a","b","c", "city":"New York"}`

For filter we can put,
`{"city":"New York"}`

For update-data we can put,
`{"city":"Washington"}`

Then all those 3 documents change into
`{"name":"a","b","c", "city":"Washington"}`


So both filter and update-data has to be of type object, for multiple update-data the user might use iterator.